### PR TITLE
JAMES-3305 Task manager deserialization error handling

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java [hunk 1]
```diff
@@ -130,11 +130,9 @@ private void consumeWorkqueue() {
 
     private Mono<Task.Result> executeTask(AcknowledgableDelivery delivery) {
         TaskId taskId = TaskId.fromString(delivery.getProperties().getHeaders().get(TASK_ID).toString());
-        return Mono.fromCallable(() -> {
-            delivery.ack();
-            return new String(delivery.getBody(), StandardCharsets.UTF_8);
-        }).flatMap(json ->
-            deserialize(json, taskId)
+        return Mono.fromCallable(() -> new String(delivery.getBody(), StandardCharsets.UTF_8))
```
Comment (chibenwa) [inline server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java:133]: Do we need to wrap bytes -> string operation in a mono?

File: server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java [hunk 2]
```diff
@@ -135,6 +144,42 @@ void givenANonDeserializableTaskItShouldBeFlaggedAsFailedAndItDoesNotPreventFoll
...
+        BasicProperties badTaskProperties = new BasicProperties.Builder()
+            .deliveryMode(PERSISTENT_TEXT_PLAIN.getDeliveryMode())
+            .priority(PERSISTENT_TEXT_PLAIN.getPriority())
+            .contentType(PERSISTENT_TEXT_PLAIN.getContentType())
+            .headers(ImmutableMap.of("taskId", badTask.getId().asString()))
+            .build();
+
+        rabbitMQExtension.getSender()
+            .send(Mono.just(new OutboundMessage(EXCHANGE_NAME,
+                ROUTING_KEY, badTaskProperties, "BAD_PAYLOAD!".getBytes(StandardCharsets.UTF_8))))
```
Comment (chibenwa) [inline server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java:161]: We deserve 2 tests, one for the bad payload, onefor the bad properties?

File: server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java [hunk 3]
```diff
@@ -135,6 +144,42 @@ void givenANonDeserializableTaskItShouldBeFlaggedAsFailedAndItDoesNotPreventFoll
...
+            .deliveryMode(PERSISTENT_TEXT_PLAIN.getDeliveryMode())
+            .priority(PERSISTENT_TEXT_PLAIN.getPriority())
+            .contentType(PERSISTENT_TEXT_PLAIN.getContentType())
+            .headers(ImmutableMap.of("taskId", TASK_WITH_ID.getId().asString()))
+            .build();
+
+        rabbitMQExtension.getSender()
+            .send(Mono.just(new OutboundMessage(EXCHANGE_NAME,
+                ROUTING_KEY, properties, serializer.serialize(TASK_WITH_ID.getTask()).getBytes(StandardCharsets.UTF_8))))
+            .block();
```
Comment (chibenwa) [inline server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java:176]: Can't we rather use a business api to send that valid json task to the worker?

Maybe testing the Worker in isolation is not whatyou want. I advise youto write the very same test at the level of the distributed task manager.

File: server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java [hunk 4]
```diff
@@ -145,7 +146,95 @@ void givenANonDeserializableTaskItShouldBeFlaggedAsFailedAndItDoesNotPreventFoll
...
+            .send(Mono.just(new OutboundMessage(EXCHANGE_NAME,
+                ROUTING_KEY, badProperties, serializer.serialize(TASK_WITH_ID.getTask()).getBytes(StandardCharsets.UTF_8))))
+            .block();
+
+        BasicProperties properties = new BasicProperties.Builder()
+            .deliveryMode(PERSISTENT_TEXT_PLAIN.getDeliveryMode())
+            .priority(PERSISTENT_TEXT_PLAIN.getPriority())
+            .contentType(PERSISTENT_TEXT_PLAIN.getContentType())
+            .headers(ImmutableMap.of("taskId", TASK_WITH_ID.getId().asString()))
+            .build();
```
Comment (Arsnael) [inline server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java:224]: can the basic working properties be extracted into a method? or would it reduce the understanding of the tests?

File: server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java [hunk 5]
```diff
@@ -129,13 +130,16 @@ private void consumeWorkqueue() {
...
-        }).flatMap(json ->
-            deserialize(json, taskId)
-                .flatMap(task -> executeOnWorker(taskId, task)));
+        return Mono.fromCallable(() -> TaskId.fromString(delivery.getProperties().getHeaders().get(TASK_ID).toString()))
+            .flatMap(taskId -> deserialize(new String(delivery.getBody(), StandardCharsets.UTF_8), taskId)
+                .doOnNext(task -> delivery.ack())
+                .flatMap(task -> executeOnWorker(taskId, task)))
+            .onErrorResume(error -> {
+                String errorMessage = String.format("Unable to process %s %s", TASK_ID, delivery.getProperties().getHeaders().get(TASK_ID));
+                LOGGER.error(errorMessage, error);
```
Comment (chibenwa) [inline server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java:139]: Use SLF4J built in message formatting please:

```
LOGGER.error("Unable to process {} {}", TASK_ID, delivery.getProperties().getHeaders().get(TASK_ID), e);
```

**Comment by Arsnael** [comment]
I think this issue should be linked to https://issues.apache.org/jira/browse/JAMES-3308 instead of JAMES-3305 :)

**Comment by chibenwa** [comment]
```
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [ERROR] Tests run: 27, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 69.736 s <<< FAILURE! - in JUnit Vintage
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [ERROR] A sharee should be able to retrieve a mailbox after sharer moved it into another mailbox  Time elapsed: 2.638 s  <<< FAILURE!
[aed35a78aacf4793775d54b3c324753c2cc53fcb] java.lang.AssertionError: 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] Expecting:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <[null]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] to contain only:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <["10063740-c72d-11ea-a48c-ada34fe00329"]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] elements not found:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <["10063740-c72d-11ea-a48c-ada34fe00329"]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] and elements not expected:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <[null]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [INFO] 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [INFO] Results:
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [INFO] 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [ERROR] Failures: 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] [ERROR]   RabbitMQMailboxSharingTest 
[aed35a78aacf4793775d54b3c324753c2cc53fcb] Expecting:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <[null]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] to contain only:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <["10063740-c72d-11ea-a48c-ada34fe00329"]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] elements not found:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <["10063740-c72d-11ea-a48c-ada34fe00329"]>
[aed35a78aacf4793775d54b3c324753c2cc53fcb] and elements not expected:
[aed35a78aacf4793775d54b3c324753c2cc53fcb]   <[null]>
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**